src/charm.py

```diff
@@ -105,228 +110,145 @@ def _on_pgbouncer_pebble_ready(self, event: PebbleReadyEvent) -> None:
-            space, one pair per line.
-        """
-        return "\n".join([f'"{username}" "{password}"' for username, password in users.items()])
-
-    def _get_users_from_charm_config(self, users: Dict[str, str] = {}) -> Dict[str, str]:
-        """Imports users from charm config and generates passwords when necessary.
-
-        When generated, passwords are hashed using md5.
-
-        Args:
-            users: an existing dictionary of usernames and passwords
-        Returns:
-            the `users` dictionary, with usernames from charm config and corresponding generated
-            passwords appended.
-        """
-        for admin_user in self.config["pgb_admin_users"].split(","):
-            if admin_user not in users or users[admin_user] is None:
-                users[admin_user] = self._hash(self._generate_password())
-        return users
-
-    def _get_userlist_from_container(self) -> Dict[str, str]:
+    def _read_userlist(self) -> Dict[str, str]:
         """Parses container userlist to a dict.
 
         This parses the userlist.txt stored on the container into a dictionary of strings, where
         the users are the keys and hashed passwords are the values.
 
         Returns:
-            users: a dictionary of usernames and passwords
+            a dictionary of usernames and passwords
         """
-        pgb_container = self.unit.get_container(self._pgbouncer_container)
-        if not pgb_container.can_connect():
-            logger.info("pgbouncer container not accessible, cannot access userlist")
-            return {}
-
         try:
-            userlist = pgb_container.pull(USERLIST_PATH).read()
+            userlist = self._read_file(USERLIST_PATH)
+            return pgb.parse_userlist(userlist)
         except FileNotFoundError:
-            # There is no existing userlist.txt file, so return an empty dict
-            return {}
+            logger.error("userlist not found")
 
-        parsed_userlist = {}
-        for line in userlist.split("\n"):
-            if line.strip() == "":
-                continue
-            # Userlist is formatted "{username}" "{password}""
-            username, password = line.replace('"', "").split(" ")
-            parsed_userlist[username] = password
+    def _reload_pgbouncer(self) -> None:
+        """Reloads pgbouncer application.
 
-        return parsed_userlist
+        Pgbouncer will not apply configuration changes without reloading, so this must be called
+        after each time config files are changed.
+        """
+        self.unit.status = MaintenanceStatus("Reloading Pgbouncer")
+        logger.info("reloading pgbouncer application")
 
-    def _generate_password(self) -> str:
-        """Generates a secure password.
+        pgb_container = self.unit.get_container(PGB)
+        if not pgb_container.can_connect():
+            self.unit.status = WaitingStatus(
+                "unable to restart pgbouncer - container not connected."
+            )
```

Comment:
This check can be removed as it's already being checked before the two calls to this method.

Reply:
removed